src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs

```diff
@@ -38,86 +38,24 @@ public static DynamicWinsockMethods GetMethods(AddressFamily addressFamily, Sock
-
         private WSARecvMsgDelegate? _recvMsg;
-        private WSARecvMsgDelegateBlocking? _recvMsgBlocking;
 
         private DynamicWinsockMethods(AddressFamily addressFamily, SocketType socketType, ProtocolType protocolType)
         {
             _addressFamily = addressFamily;
             _socketType = socketType;
             _protocolType = protocolType;
-            _lockObject = new object();
         }
 
-        public T GetDelegate<T>(SafeSocketHandle socketHandle)
-            where T : class
-        {
-            if (typeof(T) == typeof(AcceptExDelegate))
-            {
-                EnsureAcceptEx(socketHandle);
-                Debug.Assert(_acceptEx != null);
-                return (T)(object)_acceptEx;
-            }
-            else if (typeof(T) == typeof(GetAcceptExSockaddrsDelegate))
-            {
-                EnsureGetAcceptExSockaddrs(socketHandle);
-                Debug.Assert(_getAcceptExSockaddrs != null);
-                return (T)(object)_getAcceptExSockaddrs;
-            }
-            else if (typeof(T) == typeof(ConnectExDelegate))
-            {
-                EnsureConnectEx(socketHandle);
-                Debug.Assert(_connectEx != null);
-                return (T)(object)_connectEx;
-            }
-            else if (typeof(T) == typeof(DisconnectExDelegate))
-            {
-                EnsureDisconnectEx(socketHandle);
-                Debug.Assert(_disconnectEx != null);
-                return (T)(object)_disconnectEx;
-            }
-            else if (typeof(T) == typeof(DisconnectExDelegateBlocking))
-            {
-                EnsureDisconnectEx(socketHandle);
-                Debug.Assert(_disconnectExBlocking != null);
-                return (T)(object)_disconnectExBlocking;
-            }
-            else if (typeof(T) == typeof(WSARecvMsgDelegate))
-            {
-                EnsureWSARecvMsg(socketHandle);
-                Debug.Assert(_recvMsg != null);
-                return (T)(object)_recvMsg;
-            }
-            else if (typeof(T) == typeof(WSARecvMsgDelegateBlocking))
-            {
-                EnsureWSARecvMsgBlocking(socketHandle);
-                Debug.Assert(_recvMsgBlocking != null);
-                return (T)(object)_recvMsgBlocking;
-            }
-            else if (typeof(T) == typeof(TransmitPacketsDelegate))
-            {
-                EnsureTransmitPackets(socketHandle);
-                Debug.Assert(_transmitPackets != null);
-                return (T)(object)_transmitPackets;
-            }
-
-            Debug.Fail("Invalid type passed to DynamicWinsockMethods.GetDelegate");
-            return null;
-        }
-
-        // Private methods that actually load the function pointers.
-        private IntPtr LoadDynamicFunctionPointer(SafeSocketHandle socketHandle, ref Guid guid)
+        private static T CreateDelegate<T>(ref T? cache, SafeSocketHandle socketHandle, string guidString) where T: Delegate
```

Comment:
I think you should be able to make it:
```suggestion
        private static T CreateDelegate<T>([NotNull] ref T? cache, SafeSocketHandle socketHandle, string guidString) where T: Delegate
```